WordPress/gutenberg
# Show patterns in inserter for non-root level insert destinations

Fixes #25915 



## Description
Patterns are only allowed to be inserted on the root level. (as a child of Document)
This PR adds the option to insert patterns wherever it's possible.

It is based on two new selectors:
- `__experimentalGetParsedPatterns`: Parses all the patterns and memorizes the value. Thus we are only parsing patterns once per page session.
- `__experimentalGetAllowedPatterns`: To avoid corrupting the content, we are only allowing to insert patterns that contain blocks that are allowed to be inserted into our target block. This function filters the parsed patterns and returns only the patterns that can be inserted.

Since `parse` is a heavy operation, we are memorizing the parsed patterns. To avoid affecting the time it takes to open the inserter, we are calling the `__experimentalGetAllowedPatterns` function after loading the editor. This will cause the editor to take longer to load but throughout the usage of the block editor, we won't notice any performance downgrades.

This solution isn't ideal, but it gets the balls rolling.

## How has this been tested?
1. Open post editor
2. Insert `Columns`, choose 50 / 50
3. Insert `Paragraph` into the first column with some random text
4. Select `Paragraph` block
5. Open Inserter
6. Patterns tab should be visible and you should be able to insert patterns
7. Select the parent `Column` block of the `Paragraph`
8. Open inserter
9. Patterns tab should be hidden

## Screenshots 

https://user-images.githubusercontent.com/2256104/105723674-576c2300-5f27-11eb-9bed-c06c7d4194d5.mov



## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
